signalfx-async-wrapper.js

```diff
@@ -12,67 +14,87 @@ class SignalFxWrapper {
     originalContext,
     originalCallback,
     dimensions,
-    accessToken
+    disableTracing,
+    disableMetrics
   ) {
     this.originalObj = originalObj;
     this.originalFn = originalFn;
     this.originalEvent = originalEvent;
     this.originalContext = originalContext;
     this.originalCallback = originalCallback;
+    this.enableMetrics = sfxHelper.isMetricsEnabled(disableMetrics);
+    this.enableTracing = sfxHelper.isTracingEnabled(disableTracing);
 
-    sfxHelper.setAccessToken(accessToken);
-    sfxHelper.setLambdaFunctionContext(this.originalContext, dimensions);
-    sfxHelper.sendCounter('function.invocations', 1);
+    this.metricSender = sfxHelper;
+    if (!this.enableMetrics) {
+      this.metricSender = noopMetricSender;
+    }
+
+    this.execMeta = sfxHelper.getExecutionMetadata(this.originalContext);
+    sfxHelper.setDefaultDimensions(dimensions, this.execMeta);
+
+    this.metricSender.sendCounter("function.invocations", 1);
     if (coldStart) {
-      sfxHelper.sendCounter('function.cold_starts', 1);
+      this.metricSender.sendCounter("function.cold_starts", 1);
       coldStart = false;
     }
     return this;
   }
 
   async invoke() {
+    const tracer = tracing.tracer(!this.enableTracing);
+    const span = tracing.startSpan(tracer, this.originalEvent, this.execMeta);
+
     const startTime = new Date().getTime();
     let sent;
-    const sendDuration = async () => {
+    const flush = async () => {
       if (sent) {
         return;
       }
 
-      sfxHelper.sendGauge('function.duration', new Date().getTime() - startTime);
+      this.metricSender.sendGauge(
+        "function.duration",
+        new Date().getTime() - startTime
+      );
       sent = true;
 
-      await sfxHelper.waitForAllSends();
-    }
+      await Promise.all([tracing.flush(), sfxHelper.waitForAllSends()]);
+    };
 
     const callbackWrapper = (err, result) => {
-      sendDuration().then(() => {
+      flush().then(() => {
         this.originalCallback(err, result);
-      })
+      });
     };
 
-    let result;
-    try {
-      result = await this.originalFn.call(
-        this.originalObj,
-        this.originalEvent,
-        this.originalContext,
-        callbackWrapper
-      );
-
-      await sendDuration();
-    } catch (err) {
-      sfxHelper.sendCounter('function.errors', 1);
-      await sendDuration();
-
-      // should just return js exception instead of doing additional processing: https://aws.amazon.com/blogs/compute/node-js-8-10-runtime-now-available-in-aws-lambda/
-      return err;
-    }
+    return tracer.scope().activate(span, async () => {
+      try {
+        const result = await this.originalFn.call(
+          this.originalObj,
+          this.originalEvent,
+          this.originalContext,
+          callbackWrapper
+        );
 
-    return result;
+        span.finish();
+        await flush();
```

Comment:
both span.finish and flushes in a finally?

Reply:
Even after a decade of programming, I can never remember the code execution order of try .. catch .. finally :)

Reply:
Fixed.